Alpine httpd

ncopa opened the pull request:
Some notes.
- use env for gpg key
- make sure runtime and build time deps are fetched only once
- be are that there is a busybox httpd too in the PATH (maybe we should delete it?)

diff from the debian based:

````diff
diff -ru 2.4/Dockerfile 2.4-alpine/Dockerfile
--- 2.4/Dockerfile	2016-01-15 16:06:00.709343409 +0100
+++ 2.4-alpine/Dockerfile	2016-01-17 21:10:16.108196165 +0100
@@ -1,7 +1,7 @@
-FROM debian:jessie
+FROM alpine:3.3
 
 # add our user and group first to make sure their IDs get assigned consistently, regardless of whatever dependencies get added
-#RUN groupadd -r www-data && useradd -r --create-home -g www-data www-data
+RUN adduser -D www-data
 
 ENV HTTPD_PREFIX /usr/local/apache2
 ENV PATH $PATH:$HTTPD_PREFIX/bin
@@ -9,54 +9,51 @@
 	&& chown www-data:www-data "$HTTPD_PREFIX"
 WORKDIR $HTTPD_PREFIX
 
-# install httpd runtime dependencies
-# https://httpd.apache.org/docs/2.4/install.html#requirements
-RUN apt-get update \
-	&& apt-get install -y --no-install-recommends \
-		libapr1 \
-		libaprutil1 \
-		libapr1-dev \
-		libaprutil1-dev \
-		libpcre++0 \
-		libssl1.0.0 \
-	&& rm -r /var/lib/apt/lists/*
-
 # see https://httpd.apache.org/download.cgi#verify
-RUN gpg --keyserver ha.pool.sks-keyservers.net --recv-keys A93D62ECC3C8EA12DB220EC934EA76E6791485A8
-
+ENV GPG_KEY A93D62ECC3C8EA12DB220EC934EA76E6791485A8
 ENV HTTPD_VERSION 2.4.18
 ENV HTTPD_BZ2_URL https://www.apache.org/dist/httpd/httpd-$HTTPD_VERSION.tar.bz2
 
-RUN buildDeps=' \
-		ca-certificates \
+# see https://httpd.apache.org/docs/2.4/install.html#requirements
+RUN set -x \
+	&& apk add --no-cache --virtual .build-deps \
+		apr-dev \
+		apr-util-dev \
 		curl \
-		bzip2 \
 		gcc \
-		libpcre++-dev \
-		libssl-dev \
+		gnupg \
+		libc-dev \
 		make \
-	' \
-	set -x \
-	&& apt-get update \
-	&& apt-get install -y --no-install-recommends $buildDeps \
-	&& rm -r /var/lib/apt/lists/* \
+		openssl-dev \
+		pcre-dev \
+	&& gpg --keyserver ha.pool.sks-keyservers.net --recv-keys "$GPG_KEY" \
 	&& curl -SL "$HTTPD_BZ2_URL" -o httpd.tar.bz2 \
 	&& curl -SL "$HTTPD_BZ2_URL.asc" -o httpd.tar.bz2.asc \
 	&& gpg --verify httpd.tar.bz2.asc \
-	&& mkdir -p src/httpd \
-	&& tar -xvf httpd.tar.bz2 -C src/httpd --strip-components=1 \
+	&& mkdir -p src \
+	&& tar -jxvf httpd.tar.bz2 -C src \
 	&& rm httpd.tar.bz2* \
-	&& cd src/httpd \
+	&& cd src/httpd-$HTTPD_VERSION \
 	&& ./configure --enable-so --enable-ssl --prefix=$HTTPD_PREFIX --enable-mods-shared=most \
-	&& make -j"$(nproc)" \
+	&& make -j"$(getconf _NPROCESSORS_ONLN)" \
 	&& make install \
 	&& cd ../../ \
-	&& rm -r src/httpd \
+	&& rm -r src/httpd-$HTTPD_VERSION \
 	&& sed -ri ' \
 		s!^(\s*CustomLog)\s+\S+!\1 /proc/self/fd/1!g; \
 		s!^(\s*ErrorLog)\s+\S+!\1 /proc/self/fd/2!g; \
 		' /usr/local/apache2/conf/httpd.conf \
-	&& apt-get purge -y --auto-remove $buildDeps
+	&& runDeps="$( \
+		scanelf --needed --nobanner --recursive /usr/local \
+			| awk '{ gsub(/,/, "\nso:", $2); print "so:" $2 }' \
+			| sort -u \
+			| xargs -r apk info --installed \
+			| sort -u \
+	)" \
+	&& apk add --virtual .httpd-rundeps $runDeps \
+		apr-dev \
+		apr-util-dev \
+	&& apk del .build-deps
 
 COPY httpd-foreground /usr/local/bin/
 
diff -ru 2.4/httpd-foreground 2.4-alpine/httpd-foreground
--- 2.4/httpd-foreground	2016-01-15 16:06:00.709343409 +0100
+++ 2.4-alpine/httpd-foreground	2016-01-17 21:19:14.989254819 +0100
@@ -1,7 +1,7 @@
-#!/bin/bash
+#!/bin/sh
 set -e
 
 # Apache gets grumpy about PID files pre-existing
 rm -f /usr/local/apache2/logs/httpd.pid
 
-exec httpd -DFOREGROUND
+exec /usr/local/apache2/bin/httpd -DFOREGROUND
````

yosifkit commented:
Yeah, probably should remove the `busybox httpd`.  That should make the `exec /usr/local/apache2/bin/httpd` simpler too.

ncopa commented:
On a second thought, removing busybox httpd (a symlink) is risky as it will be recreated automatically if you apk add/del someghing. Better leave it and be aware of it.

ncopa commented:
Did the following fixes:
- rebased against current master
- rearranged dir structure
- merged in gpg usage fixes
- fixed generate-stackbrew-library to handle the alpine variant

tianon commented on 2.4/alpine/Dockerfile:
```diff
@@ -0,0 +1,64 @@
+WORKDIR $HTTPD_PREFIX
+
+# see https://httpd.apache.org/download.cgi#verify
+ENV GPG_KEY A93D62ECC3C8EA12DB220EC934EA76E6791485A8
```
This variable appears to be unused now. :+1:

tianon commented on 2.4/alpine/Dockerfile:
```diff
@@ -0,0 +1,64 @@
+FROM alpine:3.3
+
+# add our user and group first to make sure their IDs get assigned consistently, regardless of whatever dependencies get added
+RUN adduser -D www-data
```
Do you think we ought to hard-code this with UID/GID 82 like we did in PHP (https://github.com/docker-library/php/blob/27a370a6b7d254dd8726955a5eabd92a7562f32d/Dockerfile-alpine.template#L19-L26) to ensure we don't cause any conflicts with Alpine packages down the road?

tianon commented:
Can you elaborate a bit on why you changed `generate-stackbrew-library.sh` to be so different from the rest of the docker-library images, which mostly follow a pretty similar basic pattern?

I'm pretty happy with where the Dockerfile here is at, with just a couple minor questions. :+1:

ncopa commented:
No good reason. I just haven't read the other `generate-stackbrew-library.sh` i suppose. The general idea was to have a generic script that would find the dockerfiles in a directory tree and figure out the version and variants based on the dir path. That way we could reuse the exact same script regardless of what versions and variants the docker-library image has. And you could add another variant (or remove one) without needing to modify the `generate-stackbrew-library.sh`.

I can remove the `generate-stackbrew-library.sh` change and leave it to you.

ncopa commented:
removed the `generate-stackbrew-library.sh` change and added `www-data` user/group id fix.

andypost commented on 2.4/alpine/Dockerfile:
```diff
@@ -0,0 +1,65 @@
+FROM alpine:3.3
```
suppose make sense to use 3.4

andypost commented on 2.4/alpine/Dockerfile:
```diff
@@ -0,0 +1,65 @@
+FROM alpine:3.3
+
+# add our user and group first to make sure their IDs get assigned consistently, regardless of whatever dependencies get added
+RUN addgroup -g 82 -S www-data \
```
why 82?
maybe better to pass as ENV to allow override from outside

andypost commented:
Tested to build from your repo and got

> + export GNUPGHOME=/tmp/tmp.GKcJBe
+ gpg --keyserver ha.pool.sks-keyservers.net --recv-keys A93D62ECC3C8EA12DB220EC934EA76E6791485A8
gpg: keybox '/tmp/tmp.GKcJBe/pubring.kbx' created
gpg: keyserver receive failed: Address family not supported by protocol

andypost commented:
Fixed all comments in https://github.com/docker-library/httpd/pull/24
